### ISSUE TITLE
Better plugins support for Node API

### ISSUE THREAD

**Issue by donaldpipowitch** [eslint/eslint#9146]
**Tell us about your environment**

* **ESLint Version:** 4.4.1
* **Node Version:** 6.9
* **npm Version:**

**What parser (default, Babel-ESLint, etc.) are you using?**

typescript-eslint-parser

**Please show your full configuration:**

[My config.](https://github.com/donaldpipowitch/typescript-experiments/blob/ce536d241ea9c065ddd41eb0955b39d640e572c1/eslint/.eslintrc.json)

**What did you do? Please include the actual source code causing the issue.**

I tried to follow the current documentation to use the Node API of ESLint. [It looked like this.](https://github.com/donaldpipowitch/typescript-experiments/blob/bd2ef7253797e0586f8d5759db875c939a9c5321/eslint/eslint.js)

However it seems that custom plugin rules can't be loaded that way.

Now I use the `CLIEngine` directly as my "Node API", because it seems to cover more edge cases (like loading the plugins correctly). The code [looks like this](
https://github.com/donaldpipowitch/typescript-experiments/blob/ce536d241ea9c065ddd41eb0955b39d640e572c1/eslint/eslint.js).

**What did you expect to happen?**

The custom plugin rule should have been found/loaded correctly or the documentation should have mentioned that passing a `plugins` array in the config object isn't enough or the documentation should recommend to use the `CLIEngine` directly.

---

Thank you for the project! I hope my code snippet helps someone.

**Comment by not-an-aardvark** [comment]
Thanks for the report. I think this is working as designed -- `Linter` is only intended to lint your code with a given config and a given set of rules, without loading the config or plugins from the filesystem. On the other hand, `CLIEngine` should be used if you want to load plugins or configs from the filesystem.

Each `Linter` instance has a mutable map of defined rules, which it uses to resolve references in a config. You can use the `defineRule` method to add a rule (e.g. from a plugin).

```js
const linter = new Linter();

const customRule = context => ({
    Program(node) {
        context.report({ node, message: "some message" });
    }
});

linter.defineRule("my-custom-rule", customRule);

linter.verify("some source text", { rules: { "my-custom-rule": "error" } });
```

If you want to load all rules in a plugin, you can iterate over the plugin's `rules` object and define each of the rules on your `Linter` instance.

**Comment by donaldpipowitch** [comment]
Thanks for the explanation. I'll close this, because this is intended.
